internal/v4/auth_test.go

```diff
@@ -851,3 +807,59 @@ func (s *authSuite) TestDelegatableMacaroonWithBasicAuth(c *gc.C) {
 		ExpectStatus: http.StatusForbidden,
 	})
 }
+
+func (s *authSuite) TestGroupsForUserWithNoIdentity(c *gc.C) {
+	h := v4.New(s.store.Pool(), s.noMacaroonSrvParams)
+	groups, err := v4.GroupsForUser(h, "someone")
+	c.Assert(err, gc.IsNil)
+	c.Assert(groups, gc.HasLen, 0)
+}
+
+func (s *authSuite) TestGroupsForUserWithInvalidIdentityURL(c *gc.C) {
+	p := s.srvParams
+	p.IdentityAPIURL = ":::::"
+	h := v4.New(s.store.Pool(), p)
+	groups, err := v4.GroupsForUser(h, "someone")
+	c.Assert(err, gc.ErrorMatches, `parse :::::/v1/u/someone/idpgroups: missing protocol scheme`)
+	c.Assert(groups, gc.HasLen, 0)
+}
+
+func (s *authSuite) TestGroupsForUserWithDoFailure(c *gc.C) {
+	h := v4.New(s.store.Pool(), s.srvParams)
+	s.PatchValue(&http.DefaultClient.Transport, errorTransport("some error"))
+	groups, err := v4.GroupsForUser(h, "someone")
+	c.Assert(err, gc.ErrorMatches, `cannot get groups from http://.*/v1/u/someone/idpgroups: Get http://.*/v1/u/someone/idpgroups: some error`)
+	c.Assert(groups, gc.HasLen, 0)
+}
+
+func (s *authSuite) TestGroupsForUserWithInvalidBody(c *gc.C) {
+	h := v4.New(s.store.Pool(), s.srvParams)
+	s.idM.body = "bad"
+	groups, err := v4.GroupsForUser(h, "someone")
+	c.Assert(err, gc.ErrorMatches, `cannot unmarshal response from http://.*/v1/u/someone/idpgroups: .*`)
+	c.Assert(groups, gc.HasLen, 0)
+}
+
+func (s *authSuite) TestGroupsForUserWithErrorResponse(c *gc.C) {
+	h := v4.New(s.store.Pool(), s.srvParams)
+	s.idM.body = `{"message":"some error","code":"some code"}`
+	s.idM.status = http.StatusUnauthorized
+	groups, err := v4.GroupsForUser(h, "someone")
+	c.Assert(err, gc.ErrorMatches, `cannot get groups from http://.*/v1/u/someone/idpgroups: some error`)
+	c.Assert(groups, gc.HasLen, 0)
+}
+
+func (s *authSuite) TestGroupsForUserWithBadErrorResponse(c *gc.C) {
+	h := v4.New(s.store.Pool(), s.srvParams)
+	s.idM.body = `{"message":"some error"`
+	s.idM.status = http.StatusUnauthorized
+	groups, err := v4.GroupsForUser(h, "someone")
+	c.Assert(err, gc.ErrorMatches, `cannot unmarshal error response from http://.*/v1/u/someone/idpgroups: .*`)
+	c.Assert(groups, gc.HasLen, 0)
+}
```

Comment:
I know it's covered by other things, but an explicit test for the success case might be nice.